Hi @kingbuzzman,

Thanks for reviewing! The primary use case occurs when running **Mypy in strict mode (`mypy --strict` / `--disallow-untyped-defs`)** on downstream user codebases or type-checked pytest plugins.

### Real-world Example

Consider a Django project configured with `strict = true` in `pyproject.toml` or `mypy.ini`:

```python
# tests/test_queries.py
import pytest
from pytest_django.asserts import assertNumQueries
from myapp.models import User

def test_user_query_count() -> None:
    # 1. Context manager usage with strict typing
    with assertNumQueries(2):
        User.objects.filter(is_active=True).count()

    # 2. Callable invocation usage
    assertNumQueries(1, lambda: User.objects.get(id=1)) 

Mypy Error Output:

error: Function is missing a type annotation for parameter "func"  [no-untyped-def]
error: Function is missing a return type annotation  [no-untyped-def]

--- a/pytest_django/plugin.py
+++ b/pytest_django/plugin.py
@@ -588,7 +588,7 @@ def _django_setup_unittest(
 
     original_runtest = TestCaseFunction.runtest
 
-    def non_debugging_runtest(self) -> None:  # noqa: ANN001
+    def non_debugging_runtest(self: TestCaseFunction) -> None:
         self._testcase(result=self)
 
     from django.test import SimpleTestCase

